### DISCUSSION TITLE
How to integrate _AppStorage with TCA?

### DISCUSSION THREAD

**Discussion by tkunst891** [tgrapperon/swift-dependencies-additions#64] [Q&A] [answered]
The `@Dependency.AppStorage` implementation seems to me very helpful in combination with the UserDefaults dependency - thanks for that! But it seems not obvious to me how to use it in a TCA application.

I tried to include a `@Dependency.AppStorage` value in a reducer's State, but it does not compile, because it is not `Equatable`:
```swift
struct State: Equatable {
  @Dependency.AppStorage("mySetting") var mySetting = 100
  // ... other state
```
This will give a compile error. But it does not seem to make sense to make `AppStorage` conform to `Equatable` - I do not see how to properly define the `==` operator. There will never be two instances with differing values.

Also I do not understand how I am expected to update an `AppStorage` value in a reducer. TCA discourages executing sideeffects directly in the reduce function, and setting a `UserDefault` is a sideeffect. So you would need to use an effect (`.fireAndForget`), but such an effect can not change the state.

As a third issue it is not trivial to observe a changing `UserDefault` and react to changes in a reducer.

Anybody have suggestions or has already successfully used `_AppStorageDependency` in a TCA application?

**Comment by tgrapperon** [comment]
Hey @tkunst891! In my mind, you would use it in the reducer and use its async sequence to update a local value. The sequence always emit on "connection", so you can use it to set your initial value. So you'd produce early a `Effect` like:
```swift
.run { send in
  for await p in self.$pref {
    await send(.pref(p)) // An action that sets `pref` in your state.
  }
}
```
If you only want the value without observing it, then I agree that the side effect is not optimal. If you want to avoid it, you can also fish the value in a `.task` and emit an action that sets it. 

`DependencyAdditions` was written with vanilla SwiftUI and server-side Swift in mind. It is likely that you can write a special reducer that would provide a binding to the value. This seems interesting. I'll think about it.

**Reply by tgrapperon** [reply, marked as answer]
That being said, this feature is experimental, so suggestions are welcome!

**Reply by tkunst891** [reply]
Thanks for answering. I currently can't think of an improvement to your library which would not involve a dependency on swift-composable-architecture (which your library should of course not have, it should remain to be usable without TCA). For my own use, I have created a small helper package to let me work with @Dependency.AppStorage: https://github.com/tkunst891/dependency-appstorage-tca.git.
